apps/api/src/app/webhooks/webhooks.controller.ts

```diff
@@ -0,0 +1,83 @@
+import {
+  Body,
+  ClassSerializerInterceptor,
+  Controller,
+  NotFoundException,
+  Param,
+  Post,
+  UseInterceptors,
+} from '@nestjs/common';
+import { IntegrationEntity, MessageRepository } from '@novu/dal';
+import { ChannelTypeEnum, EmailProviderIdEnum } from '@novu/shared';
+import { MailFactory } from '../events/services/mail-service/mail.factory';
+import {
+  GetDecryptedIntegrations,
+  GetDecryptedIntegrationsCommand,
+} from '../integrations/usecases/get-decrypted-integrations';
+
+@Controller('/webhooks')
+@UseInterceptors(ClassSerializerInterceptor)
+export class WebhooksController {
+  private mailFactory = new MailFactory();
+
+  constructor(
+    private getDecryptedIntegrationsUsecase: GetDecryptedIntegrations,
+    private messageRepository: MessageRepository
+  ) {}
+
+  @Post('/:orgId/:envId/email/:providerId')
+  public async emailWebhook(
+    @Param('orgId') orgId: string,
+    @Param('envId') envId: string,
+    @Param('providerId') providerId: string,
+    @Body() body: any
+  ) {
+    const integration: IntegrationEntity = await this.getDecryptedIntegrationsUsecase.execute(
+      GetDecryptedIntegrationsCommand.create({
+        environmentId: envId,
+        organizationId: orgId,
+        providerId: providerId as EmailProviderIdEnum,
+        channelType: ChannelTypeEnum.EMAIL,
+        findOne: true,
+      })
+    )[0];
+
+    if (!integration) {
+      throw new NotFoundException(`Integration for ${providerId} was not found`);
+    }
+
+    const from = integration?.credentials.from || 'no-reply@novu.co';
+
+    const handler = this.mailFactory.getHandler(integration, from);
+    if (!handler) {
+      throw new NotFoundException(`Handler for integration of ${providerId} was not found`);
+    }
+    handler.buildProvider(integration.credentials, from);
+
+    const provider = handler.getProvider();
+    if (!provider.getMessageId || !provider.parseEventBody) {
+      throw new NotFoundException(`Provider with ${providerId} can not handle webhooks`);
+    }
+
+    const messageIdentifire = provider.getMessageId(body);
```

Comment:
I think most providers will batch events into an array, so we might need to handle it here, in case of an array or in case of a singular object. I would just treat everything as an array always.